src/embeddings.rs

```diff
@@ -298,3 +336,19 @@ where
 
     Ok(embeddings)
 }
+
+struct Skips<'a>(HashSet<&'a str>);
+
+impl<'a> FromPyObject<'a> for Skips<'a> {
```

Comment:
Did you try accepting `HashSet` as the argument? Would be nice if the method could take both Python sets and lists. (Which I assume doesn't work now with downcasts.)

Reply:
`HashSet` can't be used as argument, it doesn't implement the conversion.

All iterables can be passed as arguments now:

```Rust
        let mut set = HashSet::new();
        for el in ob.iter().map_err(|_| exceptions::TypeError::py_err("Iterable expected"))? {
            set.insert(el?.extract()?);
        }
        Ok(Skips(set))
```

Reply:
I'm slowly getting the hang of the conversion business in pyo3. 

I'm adding more verbose error messages for incompatible types, right now iterables with, e.g. ints only throw a `TypError` without any additional information.

Reply:
Seems like `ob.iter()` segfaults when we give input like:

```Python
embeddings.embedding_similarity(np.ones(300, dtype=np.float32), skip=1)
```

Reply:
I opened an issue at https://github.com/PyO3/pyo3/issues/557

Reply:
Might be useful to have a backtrace

```
$ gdb python
run my-python-script-that triggers-the-bug
[crashes]
bt
```

Even better with a debug build. For a full backtrace, replace `bt` by `bt full`.

Reply:
```
#0  0x00007fffc9bcbd5e in pyo3::ffi3::objectabstract::PyIter_Check (o=0x0) at /home/seb/.cargo/registry/src/github.com-1ecc6299db9ec823/pyo3-0.7.0/src/ffi3/objectabstract.rs:162
#1  0x00007fffc9b35062 in pyo3::types::iterator::PyIterator::from_object (py=..., obj=0x555555ecfbf0) at /home/seb/.cargo/registry/src/github.com-1ecc6299db9ec823/pyo3-0.7.0/src/types/iterator.rs:44
#2  0x00007fffc9b8f73f in <T as pyo3::objectprotocol::ObjectProtocol>::iter (self=0x555555ecfbf0) at /home/seb/.cargo/registry/src/github.com-1ecc6299db9ec823/pyo3-0.7.0/src/objectprotocol.rs:446
#3  0x00007fffc9b6eef0 in <finalfusion::embeddings::Skips as pyo3::conversion::FromPyObject>::extract (ob=0x555555ecfbf0) at src/embeddings.rs:344
#4  0x00007fffc9b98786 in <core::option::Option<T> as pyo3::conversion::FromPyObject>::extract (obj=0x555555ecfbf0) at /home/seb/.cargo/registry/src/github.com-1ecc6299db9ec823/pyo3-0.7.0/src/conversion.rs:287
#5  0x00007fffc9b8f878 in <T as pyo3::objectprotocol::ObjectProtocol>::extract (self=0x555555ecfbf0) at /home/seb/.cargo/registry/src/github.com-1ecc6299db9ec823/pyo3-0.7.0/src/objectprotocol.rs:485
#6  0x00007fffc9b5d3f4 in finalfusion::embeddings::__init16148627322195864837::__init16148627322195864837::__wrap::{{closure}} () at src/embeddings.rs:34
#7  0x00007fffc9b74861 in finalfusion::embeddings::__init16148627322195864837::__init16148627322195864837::__wrap (_slf=0x7fffca318390, _args=0x7ffff7e772b0, _kwargs=0x7ffff7f09630) at src/embeddings.rs:34
#8  0x0000555555651203 in _PyCFunction_FastCallDict (kwargs=0x7ffff7f09630, nargs=140736585630608, args=0x555555af9408, func_obj=0x7fffca32a948) at Objects/methodobject.c:231
#9  _PyCFunction_FastCallKeywords (func=func@entry=0x7fffca32a948, stack=stack@entry=0x555555af9408, nargs=nargs@entry=1, kwnames=kwnames@entry=0x7ffff7e834e0) at Objects/methodobject.c:294
#10 0x00005555556df7d4 in call_function (pp_stack=pp_stack@entry=0x7fffffffd5b0, oparg=<optimized out>, kwnames=kwnames@entry=0x7ffff7e834e0) at Python/ceval.c:4837
#11 0x00005555556e2078 in _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>) at Python/ceval.c:3351
#12 0x00005555556df3c3 in PyEval_EvalFrameEx (throwflag=0, f=0x555555af9288) at Python/ceval.c:754
#13 _PyEval_EvalCodeWithName (_co=_co@entry=0x7ffff7f06ae0, globals=globals@entry=0x7ffff7f06ae0, locals=locals@entry=0x7ffff7f642a0, args=args@entry=0x0, argcount=argcount@entry=0, kwnames=kwnames@entry=0x0, 
    kwargs=0x0, kwcount=0, kwstep=2, defs=0x0, defcount=0, kwdefs=0x0, closure=0x0, name=0x0, qualname=0x0) at Python/ceval.c:4166
#14 0x00005555556e03d3 in PyEval_EvalCodeEx (closure=0x0, kwdefs=0x0, defcount=0, defs=0x0, kwcount=0, kws=0x0, argcount=0, args=0x0, locals=locals@entry=0x7ffff7f642a0, globals=globals@entry=0x7ffff7f06ae0, 
    _co=_co@entry=0x7ffff7f06ae0) at Python/ceval.c:4187
#15 PyEval_EvalCode (co=co@entry=0x7ffff7f06ae0, globals=globals@entry=0x7ffff7f47318, locals=locals@entry=0x7ffff7f47318) at Python/ceval.c:731
#16 0x00005555555b3c4b in run_mod (arena=0x7ffff7f642a0, flags=0x7fffffffd8bc, locals=0x7ffff7f47318, globals=0x7ffff7f47318, filename=0x7ffff7e86180, mod=0x555555b62c00) at Python/pythonrun.c:1025
#17 PyRun_FileExFlags (fp=0x555555b381f0, filename_str=<optimized out>, start=<optimized out>, globals=0x7ffff7f47318, locals=0x7ffff7f47318, closeit=1, flags=0x7fffffffd8bc) at Python/pythonrun.c:978
#18 0x00005555555b3e4d in PyRun_SimpleFileExFlags (fp=0x555555b381f0, filename=<optimized out>, closeit=1, flags=0x7fffffffd8bc) at Python/pythonrun.c:419
#19 0x00005555555cb081 in run_file (p_cf=0x7fffffffd8bc, filename=0x555555a9e6f0 L"bug.py", fp=0x555555b381f0) at Modules/main.c:340
#20 Py_Main (argc=<optimized out>, argv=<optimized out>) at Modules/main.c:810
#21 0x00005555555aa25a in main (argc=2, argv=<optimized out>) at ./Programs/python.c:69
```

Confirms that the issue comes from the `PyIterator`. Interestingly, the if branch that leads to the error has a comment:
```
// this is not right, but this cause of segfault check #71
```
But 71 is fixed at pyo3.

Reply:
For now we could use the following work-around:
```Rust
        let iter = if <PySet as PyTypeInfo>::is_instance(ob) {
            ob.iter().unwrap()
        } else if <PyList as PyTypeInfo>::is_instance(ob) {
            ob.iter().unwrap()
        } else {
            return Err(exceptions::TypeError::py_err("Iterable expected"));
        };
```

This doesn't lead to segfaults and allows to pass sets and lists.

Reply:
I have submitted a PR upstream. If they are quick in handling it, we could ask them if they could cherry pick it on the 0.7 branch as well and release 0.7.1. Otherwise, I am fine with the work around, but we should put a pointer to the bug as a comment.

Reply:
I think they had been working on a fix already, although nothing changed on the branch in the last 10 days: https://github.com/PyO3/pyo3/tree/fix_494

Reply:
Whoops. Well, it was a nice expedition into pyo3 ;).

Reply:
Your PR was merged https://github.com/PyO3/pyo3/pull/558, once they release a new minor version, we should be able to merge this PR.

Reply:
It doesn't look like they are cherry-picking your fix and they're also not releasing a new minor version. 

Options would be: Wait some more, use a git version of pyo3 or use the workaround and fix later.

What do you prefer?

Reply:
Use the workaround and make an issue (pointing to this PR) so that we don't forget to change it once a new PyO3 is out.